crepecake/src/main/java/com/google/cloud/tools/crepecake/hash/CountingDigestOutputStream.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.hash;
+
+import com.google.cloud.tools.crepecake.blob.BlobDescriptor;
+import com.google.cloud.tools.crepecake.image.DescriptorDigest;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.security.DigestException;
+import java.security.DigestOutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+
+/** A {@link DigestOutputStream} that also keeps track of the total number of bytes written. */
+public class CountingDigestOutputStream extends DigestOutputStream {
+
+  private static final String SHA_256_ALGORITHM = "SHA-256";
+
+  /** Keeps track of the total number of bytes appended. */
+  private long totalBytes = 0;
+
+  /** Wraps the {@code outputStream}. */
+  public CountingDigestOutputStream(OutputStream outputStream) throws NoSuchAlgorithmException {
+    super(outputStream, null);
+    setMessageDigest(MessageDigest.getInstance(SHA_256_ALGORITHM));
```

Comment:
Peer Javadoc:

Every implementation of the Java platform is required to support the following standard MessageDigest algorithms:

MD5
SHA-1
SHA-256

Given that, I think it would be reasonable to to catch the NoSuchAlgorithmException and convert it into a RuntimeException here, since failure to support this algorithm indicates a broken VM.